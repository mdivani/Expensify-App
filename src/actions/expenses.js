import uuid from 'uuid';
import database from '../firebase/firebase';

//add expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createAt = 0
    } = {}
) => {
    return (dispatch) => {
        const expense = {description, note, amount, createAt };
        return database.ref('expenses').push(expense)
                      .then((ref) => {
                        dispatch(addExpense({
                            id: ref.key,
                            ...expense
                        }))
                      });
    }
}

//remove expense
export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    expenseId: id
})
//edit expense
export const editExpense = (id, updates) => ({
    type:'EDIT_EXPENSE',
    id,
    updates
})

//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snap) => {
            const expenses = [];
            snap.forEach(childSnap => {
                expenses.push({
                    id: childSnap.key,
                    ...childSnap.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    }
}