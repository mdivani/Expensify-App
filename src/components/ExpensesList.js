import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import SelectedExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className='content_container'>
        <div className='list-header'>
            <div className='show-for-mobile'>Expenses</div>
            <div className='show-for-desktop'>Expense</div>
            <div className='show-for-desktop'>Amount</div>
        </div>
        <div className='list-body'>
        {
            props.expenses.length === 0 ? (
                <div className='list-item list-item--message'>
                    <span>No expenses</span>                
                </div>
            )
            : (
                props.expenses.map((expense, index) => {
                    return <ExpenseListItem {...expense} key={index} />
                })
            )
        }        
        </div>
    </div>
);

const fromStateToProp = (state) => ({
    expenses: SelectedExpenses(state.expenses, state.filter)
})

export default connect(fromStateToProp)(ExpenseList);