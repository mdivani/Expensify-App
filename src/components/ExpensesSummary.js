import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectedExpenses from '../selectors/expenses';
import totalExpensesSelector from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
    const totallingExpenses = numeral(totalExpensesSelector(props.expenses)/100).format('$0,0.00');
    const totalExpenses = props.expenses.length;
return (
    <div>
        {totalExpenses > 0 && <p>
            {
            props.expenses.length === 1 ? 
            `Viewing 1 expense totalling ${totallingExpenses}` : 
            `Viewing ${totalExpenses} expenses totalling ${totallingExpenses}`
            }
        </p>
        }
</div>
)
}

const fromStateToProp = (state) => ({
    expenses: selectedExpenses(state.expenses, state.filter)
})

export default connect(fromStateToProp)(ExpensesSummary);