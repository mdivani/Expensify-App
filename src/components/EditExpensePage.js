import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

    handleEditExpense = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    handleRemoveExpense  = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense = {this.props.expense}
                    onSubmit={this.handleEditExpense} 
                />
                <button onClick={this.handleRemoveExpense}>
                    Remove
                </button>
            </div>
        )
    }
} 


const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => {
        return expense.id === props.match.params.id;
    })
});

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);