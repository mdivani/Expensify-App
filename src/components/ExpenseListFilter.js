import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null,
        sortBy: 'date'
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    
    onSortChange = (e) => {
        const sortBy = e.target.value;
            if(sortBy === 'date') {
                this.props.sortByDate();
            }
            else if(sortBy === 'amount') {
                this.props.sortByAmount();
            }
    }
    
    render() {
        return (
            <div className='content_container'>
                <div className='input-group'>
                    <div className='input-group__item'>
                        <input
                             className='text-input'
                             placeholder='Search expenses'
                             type='text' 
                             value={this.props.filter.text} 
                             onChange={this.onTextChange} />
                    </div>
                    <div className='input-group__item'>
                        <select 
                        className='select'
                        value={this.props.filter.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className='input-group__item'>
                        <DateRangePicker
                        startDate={this.props.filter.startDate}
                        endDate={this.props.filter.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                        />
                    </div>
                </div>
            </div>
        );
    }
}


const fromStateToProp = (state) => {
    return {
        filter: state.filter
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setEndDate: (date) => dispatch(setEndDate(date))
})

export default connect(fromStateToProp, mapDispatchToProps)(ExpenseListFilter);

