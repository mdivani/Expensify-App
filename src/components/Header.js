import React from 'react';
import {Link} from 'react-router-dom';
import { startLogout } from '../actions/Auth';
import { connect } from 'react-redux';

export const Header = ({startLogout}) => (
    <header className='header'>
    <div className='content_container'>
        <div className='header__content'>
            <Link className='header__title' to='/dashboard'>
                <h1 className='heading heading--brand'>
                    <span className='heading--brand__first'>Mo</span>
                    <span className='heading--brand__second'>Bill</span>
                </h1>
            </Link>
            <button className='button button--link' onClick={startLogout}>Logout</button>
        </div>    
    </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);