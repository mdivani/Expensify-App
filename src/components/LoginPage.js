import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => {
    return (
        <div className='box_layout'>
            <div className='box_layout__box'>
                <h1 className='box_layout__title'>Expensify</h1>
                <p>time to take your expenses under control.</p>
                <button className='button' onClick={startLogin}>Login with Google</button>
            </div>
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);