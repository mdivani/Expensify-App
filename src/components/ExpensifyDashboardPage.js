import React from 'react';
import ExpensesList from './ExpensesList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary';

const ExpensifyDashboardPage = () => (
    <div>
        <ExpensesSummary />    
        <ExpenseListFilter />
        <ExpensesList />
    </div>
);

export default ExpensifyDashboardPage;