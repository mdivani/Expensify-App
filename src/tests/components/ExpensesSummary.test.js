import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';


test('check amount of visible expenses with 1 expense', () => {
    const wrapper = shallow(
        <ExpensesSummary
         expenses={[expenses[0]]} 
        />);
    expect(wrapper).toMatchSnapshot();
});

test('check total of visible expenses with multiple expenses', () => {
    const wrapper = shallow(
        <ExpensesSummary
         expenses={expenses} 
        />);
    expect(wrapper).toMatchSnapshot();    
});