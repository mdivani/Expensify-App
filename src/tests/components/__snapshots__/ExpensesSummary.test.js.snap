// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check amount of visible expenses with 1 expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1.95
  </p>
</div>
`;

exports[`check total of visible expenses with multiple expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $1,141.95
  </p>
</div>
`;
